core/src/main/java/org/apache/calcite/sql/type/SqlTypeFactoryImpl.java

```diff
@@ -88,6 +90,14 @@ public SqlTypeFactoryImpl(RelDataTypeSystem typeSystem) {
     if (maxPrecision >= 0 && precision > maxPrecision) {
       precision = maxPrecision;
     }
+    if (typeName == SqlTypeName.DECIMAL) {
+      if (precision != RelDataType.PRECISION_NOT_SPECIFIED && precision <= 0) {
+        throw RESOURCE.invalidPrecisionForDecimalType(precision, maxPrecision).ex();
+      }
+      if (scale != RelDataType.SCALE_NOT_SPECIFIED && (scale < 0 || scale > precision)) {
```

Comment:
it turns our that there are SQL variants which allow both negative scales and scale > precision. There is no reason for Calcite not to support these.
See https://issues.apache.org/jira/browse/CALCITE-6407
https://issues.apache.org/jira/browse/CALCITE-6406

Reply:
Java supports negative scales. It essentially means that the number has lots of zeros not represented. E.g., 1000 is 1 with scale -3.

Reply:
I appreciate your reminder. How about we add additional configurations to control this behavior about negative scales and scale>precision？ We may need two different configurations.

Reply:
yes, these should be in the type system probably, same as rounding
I personally haven't thought too much about the rules or arithmetic on negative scales
several algorithms used by calcite (e.g., infer types for arithmetic results) may need to be revisited if this is possible. I think that the existing ones were designed assuming that scale cannot be negative.

Reply:
Yes. At first, I tested it in different databases, but it still missed some features that may have been in newer versions. I am preparing to revert the unreasonable test cases in this PR(if so, this PR only left some unit tests about CAST a literal to SECAIMAL type). Do you think this PR is still necessary? If not, I am prepared to close this PR and Jira issue.

Reply:
This PR is still useful for a type system that does not support negative scales.